models/locked_resource.go

```diff
@@ -0,0 +1,121 @@
+		return nil, fmt.Errorf("get locked resource %s:%d: %v", lockType, lockKey, err)
+	} else if !has {
+		return nil, fmt.Errorf("unexpected upsert fail  %s:%d", lockType, lockKey)
+	}
+
+	// Once active, the locked resource is tied to a specific session
+	resource.engine = e
+
+	return resource, nil
+}
+
+// UpdateValue updates the value of the counter of a locked resource
+func (r *LockedResource) UpdateValue() error {
+	// Bypass ORM to support lock_type == "" and lock_key == 0
+	_, err := r.engine.Exec("UPDATE locked_resource SET counter = ? WHERE lock_type = ? AND lock_key = ?",
+		r.Counter, r.LockType, r.LockKey)
+	return err
+}
+
+// Delete deletes the locked resource from the database,
+// but the key remains locked until the end of the transaction
+func (r *LockedResource) Delete() error {
+	// Bypass ORM to support lock_type == "" and lock_key == 0
+	_, err := r.engine.Exec("DELETE FROM locked_resource WHERE lock_type = ? AND lock_key = ?", r.LockType, r.LockKey)
+	return err
+}
+
+// DeleteLockedResourceKey deletes a locked resource by key
+func DeleteLockedResourceKey(e Engine, lockType string, lockKey int64) error {
+	// Bypass ORM to support lock_type == "" and lock_key == 0
+	_, err := e.Exec("DELETE FROM locked_resource WHERE lock_type = ? AND lock_key = ?", lockType, lockKey)
+	return err
+}
+
+// TemporarilyLockResourceKey locks the given key but does not leave a permanent record
+func TemporarilyLockResourceKey(e Engine, lockType string, lockKey int64) error {
+	// Temporary locked resources should not exist in the table.
+	// This allows us to use a simple INSERT to lock the key.
+	_, err := e.Exec("INSERT INTO locked_resource (lock_type, lock_key) VALUES (?, ?)", lockType, lockKey)
+	if err == nil {
+		_, err = e.Exec("DELETE FROM locked_resource WHERE lock_type = ? AND lock_key = ?", lockType, lockKey)
+	}
+	return err
+}
+
+// GetLockedResourceCtx gets or creates a pessimistic lock on the given type and key
+func GetLockedResourceCtx(ctx DBContext, lockType string, lockKey int64) (*LockedResource, error) {
+	return GetLockedResource(ctx.e, lockType, lockKey)
+}
+
+// DeleteLockedResourceKeyCtx deletes a locked resource by key
+func DeleteLockedResourceKeyCtx(ctx DBContext, lockType string, lockKey int64) error {
+	return DeleteLockedResourceKey(ctx.e, lockType, lockKey)
+}
+
+// TemporarilyLockResourceKeyCtx locks the given key but does not leave a permanent record
+func TemporarilyLockResourceKeyCtx(ctx DBContext, lockType string, lockKey int64) error {
+	return TemporarilyLockResourceKey(ctx.e, lockType, lockKey)
+}
+
+// upsertLockedResource will create or lock the given key in the database.
+// the function will not return until it acquires the lock or receives an error.
+func upsertLockedResource(e Engine, resource *LockedResource) (err error) {
+	// An atomic UPSERT operation (INSERT/UPDATE) is the only operation
+	// that ensures that the key is actually locked.
```

Comment:
this should go in xorm .. a atomic InsertUpdate func ... This is needed in many places

Reply:
The reason I didn't go down that path is that I believe we need pessimistic locking in quite a few places, not just this insert.